flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java

```diff
@@ -0,0 +1,380 @@
+
+    /** Changes that are not yet uploaded (upload not requested). */
+    private final NavigableMap<SequenceNumber, StateChangeSet> notUploaded = new TreeMap<>();
+
+    /** Uploaded changes, ready for use in snapshots. */
+    @GuardedBy("lock")
+    private final NavigableMap<SequenceNumber, UploadResult> uploaded = new TreeMap<>();
+
+    /**
+     * Highest {@link SequenceNumber} for which upload has failed (won't be restarted), inclusive.
+     */
+    @Nullable
+    @GuardedBy("lock")
+    private Tuple2<SequenceNumber, Throwable> highestFailed;
+
+    @GuardedBy("lock")
+    private boolean closed;
+
+    FsStateChangelogWriter(
+            UUID logId,
+            KeyGroupRange keyGroupRange,
+            StateChangeUploader uploader,
+            long preEmptivePersistThresholdInBytes,
+            FsStateChangelogCleaner cleaner) {
+        this.logId = logId;
+        this.keyGroupRange = keyGroupRange;
+        this.uploader = uploader;
+        this.preEmptivePersistThresholdInBytes = preEmptivePersistThresholdInBytes;
+        this.cleaner = cleaner;
+    }
+
+    @Override
+    public void append(int keyGroup, byte[] value) throws IOException {
+        LOG.trace("append to {}: keyGroup={} {} bytes", logId, keyGroup, value.length);
+        checkState(!closed, "%s is closed", logId);
+        activeChangeSet.add(new StateChange(keyGroup, value));
+        activeChangeSetSize += value.length;
+        if (activeChangeSetSize >= preEmptivePersistThresholdInBytes) {
+            LOG.debug(
+                    "pre-emptively flush {}Mb of appended changes to the common store",
+                    activeChangeSetSize / 1024 / 1024);
+            persistInternal(notUploaded.isEmpty() ? activeSequenceNumber : notUploaded.firstKey());
+        }
+    }
+
+    @Override
+    public SequenceNumber initialSequenceNumber() {
+        return INITIAL_SQN;
+    }
+
+    @Override
+    public SequenceNumber lastAppendedSequenceNumber() {
+        LOG.trace("query {} sqn: {}", logId, activeSequenceNumber);
+        SequenceNumber tmp = activeSequenceNumber;
+        rollover();
+        return tmp;
+    }
+
+    @Override
+    public CompletableFuture<ChangelogStateHandleStreamImpl> persist(SequenceNumber from)
+            throws IOException {
+        LOG.debug(
+                "persist {} starting from sqn {} (incl.), active sqn: {}",
+                logId,
+                from,
+                activeSequenceNumber);
+        return persistInternal(from);
+    }
+
+    private CompletableFuture<ChangelogStateHandleStreamImpl> persistInternal(SequenceNumber from)
+            throws IOException {
+        synchronized (lock) {
+            ensureCanPersist(from);
+            rollover();
+            Map<SequenceNumber, StateChangeSet> tailMap = notUploaded.tailMap(from, true);
+            Map<SequenceNumber, StateChangeSet> toUpload = new HashMap<>(tailMap);
+            tailMap.clear(); // prevent re-uploads
```

Comment:
nit: `tailMap` variable is a bit confusing given that you are calling `tailMap` method twice here. Is the the intention to get the `tailMap(from, true)` from the `notUploaded` AND remove those elements from the `notUploaded` at the same time? And the issue is that there is no method like `NavigableMap#tailMapAndRemove`? If so, maybe a `private static` helper method here in this class to do just that?

Reply:
Good idea, I'll extract `drainTailMap` method.